riscv/csrs.cc

```diff
@@ -620,20 +620,49 @@ reg_t misa_csr_t::dependency(const reg_t val, const char feature, const char dep
 }
 
 bool misa_csr_t::unlogged_write(const reg_t val) noexcept {
-  // the write is ignored if increasing IALIGN would misalign the PC
-  if (!(val & (1L << ('C' - 'A'))) && (state->pc & 2))
-    return false;
-
   reg_t adjusted_val = val;
   adjusted_val = dependency(adjusted_val, 'D', 'F');
   adjusted_val = dependency(adjusted_val, 'Q', 'D');
   adjusted_val = dependency(adjusted_val, 'V', 'D');
 
   const reg_t old_misa = read();
-  const bool prev_h = old_misa & (1L << ('H' - 'A'));
   const reg_t new_misa = (adjusted_val & write_mask) | (old_misa & ~write_mask);
-  const bool new_h = new_misa & (1L << ('H' - 'A'));
 
+  const bool prev_c = old_misa & (1L << ('C' - 'A'));
+  const bool new_c = new_misa & (1L << ('C' - 'A'));
+  if (!new_c && prev_c) {
+    if (!(proc->extension_enabled(EXT_ZCB) ||
+          proc->extension_enabled(EXT_ZCMP) ||
+          proc->extension_enabled(EXT_ZCMT))) {
+      // the write is ignored if increasing IALIGN would misalign the PC
+      if (state->pc & 2)
+        return false;
+      proc->disable_extension(EXT_ZCA);
+      proc->disable_extension(EXT_ZCF);
+      proc->disable_extension(EXT_ZCD);
+    }
+  } else if (new_c && !prev_c) {
+    proc->enable_extension(EXT_ZCA);
+  }
```

Comment:
Does it make sense to disable 3 extensions but only re-enable one of the three?

Reply:
Yeah. I attempted re-enable them in following code. But something wrong in their logic  I'll fix it later.